### PR TITLE
batched seq2seq: Compute attention as matrix ops instead of for loops

### PR DESCRIPTION
**Pull request by juditacs**
In `class Attn` I replaced the `forward` function with:

~~~ python
    def forward(self, hidden, encoder_outputs):
        energy = self.attn(encoder_outputs).transpose(0, 1)  # S X B X N -> B X S X N
        e = energy.bmm(hidden.transpose(0, 1).transpose(1, 2)) # B X S X 1
        attn_energies = e.squeeze(2)  # B X S
        return F.softmax(attn_energies, 1).unsqueeze(1)  # B X 1 X S
~~~

I experienced massive speed-up (up to 100 times).

### CODE REVIEW

**Comment by ZedYeung** [comment]
The code work for all three method?

**Comment by juditacs** [comment]
By all three, do you mean dot, general and concat? Yes, it works for all three with Luong's attention and it yields the same results as the iterative version.

Bahdanau attention doesn't work at all. I filed and issue (#82) and I will try to work it out when I have some time.

**Comment by wojtekcz** [comment]
> By all three, do you mean dot, general and concat? Yes, it works for all three with Luong's attention

It doesn't work with dot as there is no self.attn variable defined in Attn.__init__()

**Comment by juditacs** [comment]
Yes, I realized that later. I'll close this PR for now and maybe fix it later when I have time.
